### PR TITLE
Add min required confirmations to new direct request logic

### PR DESCRIPTION
**Pull request by PiotrTrzpil**
The new direct request is now using LogBroadcaster, but it does not have ability to set NumConfirmations per listener yet.

Hence we need to add ability to direct request to wait certain num confirmations before starting the pipeline job, similar to how it worked in the old job runner:

```
run.TaskRuns[i].MinRequiredIncomingConfirmations = clnull.Uint32From(
            utils.MaxUint32(
                config.MinIncomingConfirmations(),
                task.MinRequiredIncomingConfirmations.Uint32,
                adapter.MinConfs()),
        )
```

Also adds a numConfirmations parameter to the direct request spec

### CODE REVIEW

File: core/services/directrequest/delegate.go
Comment (spooktheducks) [inline core/services/directrequest/delegate.go:78]: Probably ought to make this configurable, given how much we're experimenting with new chains

File: core/services/directrequest/delegate.go
Comment (spooktheducks) [inline core/services/directrequest/delegate.go:130]: Maybe rename to `unsubscribeHeads` while you're at it?

File: core/services/directrequest/delegate.go
Comment (spooktheducks) [inline core/services/directrequest/delegate.go:141]: There's a better version of this. Check out `StartOnce` and `StopOnce` on the same type. They're safer against races, especially in the `Start` case. Probably should remove `OkayToStart/Stop`

File: core/services/directrequest/delegate.go
Comment (spooktheducks) [inline core/services/directrequest/delegate.go:114]: Missing a `StartOnce`?

File: core/services/directrequest/delegate.go
Comment (spooktheducks) [inline core/services/directrequest/delegate.go:187]: What happens to things that are old enough but have been reorged away?

File: core/services/directrequest/delegate.go
Comment (samsondav) [reply core/services/directrequest/delegate.go:78]: Agree ^^. Also, do we log loudly if this is exceeded? It would be a pretty serious problem since we'd be dropping jobs.

File: core/services/directrequest/delegate.go
Comment (samsondav) [inline core/services/directrequest/delegate.go:152]: Does this force l.run to exit before Stop exits?

File: core/services/directrequest/delegate.go
Comment (samsondav) [reply core/services/directrequest/delegate.go:187]: Yup, while it is < minConfirmations we should have re-org protec in the log broadcaster.

After the log is delivered that responsibility moves to the consumer.

File: core/services/directrequest/delegate.go
Comment (samsondav) [inline core/services/directrequest/delegate.go:195]: I think it would actually be reasonable to panic here

File: core/services/directrequest/delegate.go
Comment (samsondav) [inline core/services/directrequest/delegate.go:241]: again, I think its reasonable to panic here

File: core/store/migrations/0022_add_confirmations_to_direct_request.go
Comment (samsondav) [inline core/store/migrations/0022_add_confirmations_to_direct_request.go:10]: How about leaving default NULL and falling back to the env var `MIN_INCOMING_CONFIRMATIONS` if not set?

File: core/services/directrequest/delegate.go
Comment (samsondav) [inline core/services/directrequest/delegate.go:128]: Perhaps l.run() also needs to be added to the shutdownWaitGroup?

**Review by samsondav — DISMISSED** [review]

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:130]: 👍🏼

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:114]: added

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:141]: ah yes, keepers used OkayTo.. - ok changed and added a story: https://app.clubhouse.io/chainlinklabs/story/7237/replace-oktostart-and-oktostop-calls-with-startonce-stoponce

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:187]:
```
What happens to things that are old enough but have been reorged away?
```

hmm yeah before the LB reachitecture is merged, we don't have reorg protection. Do you think we should for now add some basic protection here, like a check for IsInChain? Or wait for the merge of the other PR?

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:128]: yes, good point

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:241]: ok

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:195]: ok

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:152]: no, i thought i needed it in a test - removed now

File: core/store/migrations/0022_add_confirmations_to_direct_request.go
Comment (PiotrTrzpil) [reply core/store/migrations/0022_add_confirmations_to_direct_request.go:10]: yes, changed that to NULL now, so the following should still work fine:

```
minConfirmations := d.config.MinRequiredOutgoingConfirmations()
	if concreteSpec.NumConfirmations > minConfirmations {
		minConfirmations = concreteSpec.NumConfirmations
	}
```
as in case of NULL, concreteSpec.NumConfirmations will be 0 as far as i understand how gorm works

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go]: hm yeah - i added logging in this file on Deliver, and added a story to do it in other places https://app.clubhouse.io/chainlinklabs/story/7236/mailbox-improvements
Or do you think it should be done here? 

As for the config, i created MAILBOX_CAPACITY, but maybe you meant something more log-specific?

File: core/services/directrequest/delegate.go
Comment (samsondav) [reply core/services/directrequest/delegate.go:201]: I'd say its ok to ship this in a forthcoming PR, as long as it doesn't get forgotten

File: core/store/migrations/0022_add_confirmations_to_direct_request.go
Comment (samsondav) [reply core/store/migrations/0022_add_confirmations_to_direct_request.go]: Hmm. If it's nullable we should use `null.Int64` or something like that.

File: core/services/directrequest/delegate.go
Comment (samsondav) [inline core/services/directrequest/delegate.go:160]: commented code?

File: core/services/directrequest/delegate.go
Comment (samsondav) [inline core/services/directrequest/delegate.go:86]: What failure scenario would cause us to exceed this mailbox limit?

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:160]: oops

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:86]: in the current implementation here - it would start skipping logs if there were too many relevant logs for the same job (> MailboxCapacity) in each block.
However, that will change in the new LB implementation - LB is able to keep all logs in memory while they wait to be confirmed

Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:86]: added a comment in the code

File: core/store/migrations/0022_add_confirmations_to_direct_request.go
Comment (PiotrTrzpil) [reply core/store/migrations/0022_add_confirmations_to_direct_request.go]: ok used null.Uint32

File: core/services/directrequest/delegate.go
Comment (PiotrTrzpil) [reply core/services/directrequest/delegate.go:201]: 👍🏼

File: core/store/orm/schema.go
Comment (samsondav) [inline core/store/orm/schema.go:98]: Nit: Can we call it something more descriptive, like `DIRECT_REQUEST_LOG_BUFFER` or something?

**Review by samsondav — APPROVED** [review]

File: core/store/orm/schema.go
Comment (PiotrTrzpil) [reply core/store/orm/schema.go:98]: i thought it would be more general, but yeah

File: core/services/directrequest/delegate.go
Comment (samsondav) [inline core/services/directrequest/delegate.go:194]: This is racy. Should be added before spawning.

**Review by samsondav — CHANGES_REQUESTED** [review]

**Review by samsondav — APPROVED** [review]

**Pull request merged**
